### DISCUSSION TITLE
Looking for advice to implement a loop in Sleigh

### DISCUSSION THREAD

**Discussion by esaulenka** [NationalSecurityAgency/ghidra#7292] [Q&A] [answered]
I'm trying to improve RH850 support, in particular PUSHSP / POPSP.

These opcodes used to put (get) multiple registers in the stack in a time.

Specification says:
![image](https://github.com/user-attachments/assets/69178851-3d79-415c-af84-18ac17f22471)
`GR[cur]` here is a general register, r0..r31.

But as I know, getting the register address in Sleigh not the best approach, sometimes Ghidra produses weird decompilation in this case.

Any ideas, how to handle it? Make a context register 'SPcopy' and load/save regsters in relation to it (it will require 32 subtables both for pop and push) ? Or there is some easier way?

Full specification: [RH850G4MH User’s Manual: Software](https://www.renesas.com/en/document/mas/rh850g4mh-users-manual-software?r=1546621) pages 135, 140

**Comment by GhidorahRex** [comment, marked as answer]
I think if you have a context value that can track the current register and compare it to the end of the list, you should be able to build it up.

See buildVldmSdList in `ARMneon.sinc`. It uses a recursive sub-constructor and a 32-element subtable that is used for both push/pop (`SReg`) that should do exactly what you need here.

**Comment by esaulenka** [comment]
`vldmia` looks exactly the same as I'm looking for.
slaspec in this place is not simple enough, but i'll try to dig there.

Thanks!

**Comment by esaulenka** [comment]
Looks like it works:
https://github.com/esaulenka/ghidra_v850/commit/948ee89c882812de11f78e9cc841c8857ad4e6c1

thanks one more time!
